gtk4/src/subclass/frame.rs

```diff
@@ -0,0 +1,52 @@
+use glib::subclass::prelude::*;
+use glib::translate::*;
+use glib::Cast;
+
+use super::widget::WidgetImpl;
+use crate::{Allocation, Frame, Widget};
+
+pub trait FrameImpl: FrameImplExt + WidgetImpl {
+    fn compute_child_allocation(&self, frame: &Self::Type, allocation: &mut Allocation) {
+        self.parent_compute_child_allocation(frame, allocation)
+    }
+}
+
+pub trait FrameImplExt: ObjectSubclass {
+    fn parent_compute_child_allocation(&self, frame: &Self::Type, allocation: &mut Allocation);
+}
+
+impl<T: FrameImpl> FrameImplExt for T {
+    fn parent_compute_child_allocation(&self, frame: &Self::Type, allocation: &mut Allocation) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut ffi::GtkFrameClass;
+            let f = (*parent_class)
+                .compute_child_allocation
+                .expect("No parent class impl for \"compute_child_allocation\"");
+            f(
+                frame.unsafe_cast_ref::<Frame>().to_glib_none().0,
+                allocation.to_glib_none_mut().0,
+            )
+        }
+    }
+}
+
+unsafe impl<T: FrameImpl> IsSubclassable<T> for Frame {
+    fn override_vfuncs(class: &mut glib::Class<Self>) {
+        <Widget as IsSubclassable<T>>::override_vfuncs(class);
+
+        let klass = class.as_mut();
+        klass.compute_child_allocation = Some(frame_compute_child_allocation::<T>);
+    }
+}
+
+unsafe extern "C" fn frame_compute_child_allocation<T: FrameImpl>(
+    ptr: *mut ffi::GtkFrame,
+    allocationptr: *mut ffi::GtkAllocation,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<Frame> = from_glib_borrow(ptr);
+
+    imp.compute_child_allocation(wrap.unsafe_cast_ref(), &mut from_glib_none(allocationptr));
```

Comment:
Is there always an initialized allocation passed in here by the C callers? Or are we supposed to create one?

Reply:
GtkFrame is default implementation directly assign the values to the `Allocation` 
```c
static void
gtk_frame_real_compute_child_allocation (GtkFrame      *frame,
                                         GtkAllocation *child_allocation)
{
  GtkFramePrivate *priv = gtk_frame_get_instance_private (frame);
  int frame_width, frame_height;
  int height;

  frame_width = gtk_widget_get_width (GTK_WIDGET (frame));
  frame_height = gtk_widget_get_height (GTK_WIDGET (frame));

  if (priv->label_widget)
    {
      int nat_width, width;

      gtk_widget_measure (priv->label_widget, GTK_ORIENTATION_HORIZONTAL, -1,
                          NULL, &nat_width, NULL, NULL);
      width = MIN (frame_width, nat_width);
      gtk_widget_measure (priv->label_widget, GTK_ORIENTATION_VERTICAL, width,
                          &height, NULL, NULL, NULL);
    }
  else
    height = 0;

  child_allocation->x = 0;
  child_allocation->y = height;
  child_allocation->width = MAX (1, frame_width);
  child_allocation->height = MAX (1, frame_height - height);
}
```
So I guess the passed value is always going to be an initialized allocation unless I'm misunderstanding something

Reply:
It's allocated memory but it might have random uninitialized values inside it. It's UB in Rust to have a reference to uninitialized memory (a raw pointer is fine).

If the callers in C always initialize the allocation with something reasonable then what you do is fine. Otherwise you need to change it so that an `Allocation` is returned and directly stored in the pointer.

What can also help to understand what is the case here is to check if there are implementations of the virtual method that only read and adjust the allocation passed in. If such implementations exist then your current approach is correct.

Reply:
The thing is that I can't find anything that uses it in any parts of the code other in gtkframe.c itself which makes understanding what to expect harder. I guess creating an empty allocation with (0, 0, 1, 1) as initial values and passing it to function would work to avoid some random values?

Reply:
```cpp
static void
gtk_frame_size_allocate (GtkWidget *widget,
                         int        width,
                         int        height,
                         int        baseline)
{
  GtkFrame *frame = GTK_FRAME (widget);
  GtkFramePrivate *priv = gtk_frame_get_instance_private (frame);
  GtkAllocation new_allocation;   <----------- uninitialized

  gtk_frame_compute_child_allocation (frame, &new_allocation);
```

So I think this should simply return an `Allocation`, not pass mutable references around.

Reply:
That actually makes more sense. Thanks a lot for the help! should be ready now